flink-python/src/main/java/org/apache/flink/python/ProcessEnvironmentManager.java

```diff
@@ -0,0 +1,361 @@
+	public static final String PYTHON_REQUIREMENTS_INSTALL_DIR = "_PYTHON_REQUIREMENTS_INSTALL_DIR";
+	public static final String PYTHON_WORKING_DIR = "_PYTHON_WORKING_DIR";
+
+	private static final String PYTHON_TMP_DIR_PREFIX = "python_dist_";
+	static final String PYTHON_REQUIREMENTS_DIR = "python_requirements_target";
+	static final String PYTHON_ARCHIVES_DIR = "python_archives";
+	static final String PYTHON_FILES_DIR = "python_files_dir";
+
+	private final PythonDependencyManager dependencyManager;
+	private final String baseDirectory;
+	private final String requirementsDirectory;
+	private final String archivesDirectory;
+	private final String filesDirectory;
+	private final Map<String, String> systemEnv;
+	@Nullable private final String logDirectory;
+	private Thread shutdownHook;
+
+	private ProcessEnvironmentManager(
+		PythonDependencyManager dependencyManager,
+		String baseDirectory,
+		String filesDirectory,
+		String archivesDirectory,
+		String requirementsDirectory,
+		@Nullable String logDirectory,
+		Map<String, String> systemEnv) {
+		this.dependencyManager = dependencyManager;
+		this.baseDirectory = baseDirectory;
+		this.filesDirectory = filesDirectory;
+		this.archivesDirectory = archivesDirectory;
+		this.requirementsDirectory = requirementsDirectory;
+		this.logDirectory = logDirectory;
+		this.systemEnv = systemEnv;
+	}
+
+	@Override
+	public void open() throws Exception {
+		File baseDirectoryFile = new File(baseDirectory);
+		if (!baseDirectoryFile.exists() && !baseDirectoryFile.mkdir()) {
+			throw new IOException(
+				"Can not create the base directory: " + baseDirectory + "of python process environment manager!");
+		}
+		shutdownHook = new Thread(() -> FileUtils.deleteDirectoryQuietly(new File(baseDirectory)));
+		ShutdownHookUtil.addShutdownHookThread(shutdownHook, ProcessEnvironmentManager.class.getSimpleName(), LOG);
+	}
+
+	@Override
+	public void close() {
+		if (shutdownHook != null) {
+			shutdownHook.run();
+			ShutdownHookUtil.removeShutdownHook(shutdownHook, ProcessEnvironmentManager.class.getSimpleName(), LOG);
+			shutdownHook = null;
+		}
+	}
+
+	@Override
+	public RunnerApi.Environment createEnvironment() throws IOException {
+		prepareWorkingDir();
+		Map<String, String> generatedEnv = generateEnvironmentVariables();
+		String pythonWorkerCommand = String.join(File.separator, baseDirectory, "pyflink-udf-runner.sh");
+
+		return Environments.createProcessEnvironment(
+			"",
+			"",
+			pythonWorkerCommand,
+			generatedEnv);
+	}
+
+	/**
+	 * Just return a empty RetrievalToken because no files will be transmit via ArtifactService in process mode.
+	 *
+	 * @return The path of empty RetrievalToken.
+	 */
+	@Override
+	public String createRetrievalToken() throws IOException {
+		File retrievalToken = new File(baseDirectory,
+			"retrieval_token_" + UUID.randomUUID().toString() + ".json");
+		if (!retrievalToken.getParentFile().exists() && !retrievalToken.getParentFile().mkdir()) {
```

Comment:
It seems that this check is not necessary any more as the parent directory will be created when creating the baseDirectory.